### ISSUE TITLE
Make Holoviews docstrings meaningful for users

### ISSUE THREAD

**Issue by MarcSkovMadsen** [holoviz/holoviews#5361]
I am a user of Panel and sometimes HoloViews. I have a very hard time understanding HoloViews. It might be my own fault as I might be able to study the documentation. But that is not the way I work. I am working in my editor VS Code and use the intelligense/ tooltips and tab completion. I expect to have useful information available in my editor.

I find HoloViews very, very difficult to understand. The docstrings are not helpful at all. They seem to be written for and by the developers. Not the users of the framework.

## Code Example

I am trying to help a user here https://discourse.holoviz.org/t/cannot-delete-boxedit-after-adding/3985.

```python
import skimage
import numpy as np
import holoviews as hv

from holoviews import opts
from holoviews import streams
hv.extension('bokeh')
data = skimage.data.brain()
ds = hv.Dataset(
    (np.arange(256), np.arange(256), np.arange(10), data),
    ["x", "y", "num"],
    "count",
)
polys = hv.Polygons([])
box_stream = streams.BoxEdit(source=polys)

def stats(data):
    if not data or not any(len(d) for d in data.values()):
        return hv.Text(128, 128, "N/A")
    
    x0, x1, y0, y1 = data['x0'], data['x1'], data['y0'], data['y1']
    return hv.Text(128, 128, f"x0={x0}\nx1={x1}\ny0={y0}\ny1={y1}")

dmap = hv.DynamicMap(stats, streams=[box_stream])

im = ds.to(hv.Image, ['x', 'y'], dynamic=True)
ds.to()

plot = (im * polys + dmap).opts(
    opts.Curve(width=400, framewise=True), 
    opts.Polygons(fill_alpha=0.2, line_color='red'), 
    opts.VLine(color='black')
)
```

### hv.Dataset

The docstring is simply incomprehensible. It does not answer the most simple questions like 

- What is a Holoviews Dataset?
- Why do I need it?
- How do I use it?

It does not provide a reference example or a quick link to the documentation. I cannot use this for anything. And the `hv.Dataset` is the starting point for most things in HoloViews !!!
 
![image](https://user-images.githubusercontent.com/42288570/180501617-5002ec5a-5025-4a4f-8de9-a25fd35156d5.png)

### hv.Polygons

Wtf. It gives me the exact same docstring !!!

![image](https://user-images.githubusercontent.com/42288570/180502033-c0bb6707-80dd-4e17-9d16-0c7b3577b965.png)

### hv.DynamicMap

Wtf. It gives me the exact same docstring !!!

![image](https://user-images.githubusercontent.com/42288570/180503364-add885d5-d0a9-4e9a-9592-687887601799.png)

### BoxEdit

There is no description of what I can use BoxEdit for. There is no description of any of the arguments I can see !!!

![image](https://user-images.githubusercontent.com/42288570/180502253-8dcf6be1-a8ba-4370-aeb9-2a55761e7706.png)

### Dataset.to

What is this? How do I use it?

![image](https://user-images.githubusercontent.com/42288570/180502518-daf078ad-11d3-469f-92f1-1f890f87a580.png)

### opts

This is the most useful docstring in the example. But it would be nice if it assumed even a bit less of the user. What is an option? And why are options separate?

![image](https://user-images.githubusercontent.com/42288570/180502975-5b9ec7a0-2c5b-44fe-9b57-5ffd92a9e5fe.png)

### opts.Curve

Thanks for nothing :-). When I see something so empty I am in doubt if I am looking at a class that is not important/ should not be used.

![image](https://user-images.githubusercontent.com/42288570/180503116-ed7f5516-c20e-44ab-ad9c-7ee8cb8b8163.png)

**Comment by MarcSkovMadsen** [comment]
Another example is the `Dataset.range`. The docstring is there but needs more love for non-super users of the framework.

For example the `dimension` is this a number or a string name? More text and type annotations would be super useful. An example would be super useful. 

If I have to find lookup things in the documentation, it gets so slow and gives me mental overhead.

![image](https://user-images.githubusercontent.com/42288570/180506988-0bebcaf9-eda2-457b-b69a-9c06b7442ecb.png)

**Comment by jbednar** [comment]
Adding docstrings more consistently to both HoloViews and Panel would be an excellent project for someone to take on. It's a big job, though; it would be nice to have some idea of the priority items that would help the most. The above seems like a few scattered items; is there some way to list the level where you have the most questions or the specific set of classes or files or functions that need the most love? Without that it's hard to get started.

> It gives me the exact same docstring
I assume that means the docstring is missing and is inherited from a superclass.

**Comment by MarcSkovMadsen** [comment]
I spent maybe two 2 days on adding docstrings to Panel. And at least for me the experience of using Panel totally changed. Much easier to use. Much less mental overhead. So much more fun.

![image](https://user-images.githubusercontent.com/42288570/180573595-b4f007ad-906d-4641-ad41-fd78849d9fc5.png)

![image](https://user-images.githubusercontent.com/42288570/180573679-77ab7495-7747-4e13-9d51-7bd08f70c61f.png)

I don't think this is an overwhelming effort to do for HoloViews. The documentation is probably there in the documentation notebooks, in other docstrings.

**Comment by jbednar** [comment]
Yes, for the concrete classes like Curve the docs are in notebooks, and I'd welcome PRs that express basic concepts from the notebooks as docstrings.  The notebooks won't document abstract classes, but I guess there can't be _that_ many abstract classes, so maybe @jlstevens could take a pass on those.

**Comment by MarcSkovMadsen** [comment]
The end user would not care about the documentation for the abstract classes. What matters is the documentation of the classes they use.

**Comment by jbednar** [comment]
Well, your other issues today were about abstract class docstrings, so...

**Comment by MarcSkovMadsen** [comment]
Nope :-) The problem is that concrete classes don't have useful docstrings. Instead they had the docstring of an abstract base class.

**Comment by jbednar** [comment]
`Dataset` (above) is an example of a base class that you're requesting docs for that needs an internal maintainer like @jlstevens to document; the documentation scattered in existing notebooks won't directly cover what it does.

**Comment by MarcSkovMadsen** [comment]
I've made a few example PRs to illustrate how I would do this and how useful it is. Please give me some feedback there. Thanks.

**Comment by maximlt** [comment]
As far as I know `Dataset` isn't an abstract class.

I agree that having better docstrings is always a good idea. What you did with Panel docstrings, i.e. embedding a link to the Reference Gallery page could also be done for HoloViews, I've found that useful.

> It might be my own fault as I might be able to study the documentation. But that is not the way I work. I am working in my editor VS Code and use the intelligense/ tooltips and tab completion. I expect to have useful information available in my editor.

However, I might be wrong but you seem to say that the way you work is never to rely on any documentation website but only on what you can get in VS Code. I believe this is very unusual, and even if tooltips can prove practical in an editor like VSCode, there is no way you can get as much information in a tooltip compare to with a proper website (tooltips can't be that long otherwise they're unpractical, you can't search tooltips, etc.). Also, HoloViews is a library for visually exploring data, and documenting that is best done on a website (maybe one day you'll be able to get interactive Bokeh plot in a VSCode tooltip, not yet though). Lastly I have no data to back up this but my guess is that most users of HoloViews are in a Jupyter Notebook, which may not be the case for Panel. So docstrings changes made to HoloViews shouldn't make the experience of notebooks users worse.

I'd be interesting to investigate what docstring experience you get in VSCode with other viz libraries like seaborn and altair (7M downloads/month on PyPi).

**Comment by jbednar** [comment]
Right; Dataset is a base class, but not an abstract base class. It's useful as a concrete class that encapsulates certain functionality, and also important as the base class that implements that functionality for Elements. Thus many people need to know about it, and it's mentioned in the docs in various places, but if someone needs to know *all* about it, they go to the docstring and find almost nothing, and there's no reference manual notebook about it since it's not an Element or Layout. So it falls in a hole between our notebooks (all covering user-relevant objects) and abstract base classes (not needing notebooks since most people don't need to know about them).

I agree that the website is an important way to explore a graphical system like HoloViews, but I do sympathize with Marc's quest to have information right at his fingertips in an editor, just as I'd like to have such information right at my fingertips in Jupyter. I think having an informative docstring addresses both needs. It's true that in Jupyter we can probably handle much longer and richer docstrings, but no matter what, the current 3-sentence docstring is insufficient for such an important class.

**Comment by jlstevens** [comment]
Looking at Marc's examples, I do think the first `Dataset` example will always be a little confusing to explain due to its generality. That doesn't mean I don't think the docstring can't be improved, it can definitely be made more user friendly (and less developer focused).

I definitely think `Polygons` and other elements should have nice, consistent and user friendly docstrings. `opts` can be improved as you point out but things like `opts.Curve` are more tricky as they are dynamically generated based on the available elements.

In short, I think there are some easy wins in terms of docstring improvements in those examples screenshots (should be a relatively easy PR!), but some are a definitely trickier to implement.

**Comment by maximlt** [comment]
An easy win is having a link to the reference page in the docstring, this has been added by Marc to Panel and has proven to me pretty useful, Simon can confirm.

**Comment by MarcSkovMadsen** [comment]
I would be really happy to contribute links in docstrings to reference notebooks. For me they are a game changer in productivity when using Panel. Its such a joy to write some code and quickly jump to the reference example for more info.

**Comment by MarcSkovMadsen** [comment]
Here is another example where I trying to help a user https://discourse.holoviz.org/t/best-practices-for-interactive-layouts/4015

This use case would be similar to me reviewing or having to work with code created by colleague where I am not familiar with the framework or code used. In my work life we don't have the time to go read some web lengthy web documentation if we are just trying to change something slightly or fix a small bug.

The docstrings and type annotations of Holoviews does almost nothing to help me.

## Dataset

Hmm. What is a dataset?

![image](https://user-images.githubusercontent.com/42288570/181822489-bfaca809-bda2-4d3e-bf5e-f39c9543c23f.png)


## `.select`

Its relatively well documented. There is no example for categorical/ text data. That would have been nice. But ok

![image](https://user-images.githubusercontent.com/42288570/181822823-421fb730-83fb-4486-9129-6877dc77669f.png)

## `.to`

My guess before seeing the docstring was that I will get a`hv.Curve` with `kdims=_time` and `vdims=sig_value`. But after reading the tooltip I'm not so sure any more.

![image](https://user-images.githubusercontent.com/42288570/181823137-41b904aa-080e-442a-9943-42d62b24b8ec.png)

## `.overlay` and `.collate`

A tooltip would maybe have told me what this does in seconds. Now I have to spend 1-10 minutes googling and looking up documentation.

![image](https://user-images.githubusercontent.com/42288570/181823418-d0a9a9d3-1c78-4dfc-a976-a551387c77d6.png)

![image](https://user-images.githubusercontent.com/42288570/181823017-a73fb60a-d52b-4aa4-a70a-013ae766cba9.png)

## Additional Comment

It is very, very important for productivity that you are able to easily read the code.

*Code is read more often than it is written*. Guido Van Rossum.

**Comment by MarcSkovMadsen** [comment]
I tried looking at some Plotly examples. The deeply nested code I tested did have useful docstrings and type annotations.

**Comment by MarcSkovMadsen** [comment]
I looked at Altair. Its docstrings is of higher quality than Holoviews. 

But it also has issues as it does not document its arguments in docstrings. And its type annotations are not perfect - Their api should be simple to annotate and provide tooltips for as they very often return `self` or `self.copy(deep=True)`.

**Comment by CMRicketts** [comment]
I would like to assist with this ticket, if possible. thanks.

**Comment by MarcSkovMadsen** [comment]
Hi @CMRicketts 

Welcome to HoloViz. Sounds great. While waiting for suggestions from @philippjfr my suggestion would be that 

- You review the work in progress PR https://github.com/holoviz/hvplot/pull/827 with the purpose of aligning on the way we do docstrings.
- Then you create a new similar work in progress PR for HoloViews and start to document `hv.Curve`, `hv.Area`, .... in a similar fashion. I might join later and document some when I've finished the other one.

But its up to up :-) Find something that you are really motivated to improve and start working on it. Do a small PR initially to get feedback and alignment. Then you can make a bigger PR.
